apps/emqx_ds_backends/test/emqx_ds_backends_SUITE.erl

```diff
@@ -2297,6 +2297,103 @@ t_29_tx_latest_generation_race_condition(Config) ->
         []
     ).
 
+%% This testcase verifies that transactions that contain write
+%% operations and dirty appends are rejected when the DB reaches the
+%% storage quota. Then it verifies that the system recovers once the
+%% quota is increased.
+t_30_quotas(Config) ->
+    DB = ?FUNCTION_NAME,
+    %% This testcase uses the default implicit group created by DS:
+    Group = DB,
+    Opts = #{backend := Backend} = opts(Config),
+    ?check_trace(
+        #{timetrap => 15_000},
+        begin
+            ?assertMatch(ok, emqx_ds_open_db(DB, Opts)),
+            TxOpts = #{
+                db => DB,
+                shard => emqx_ds:shard_of(DB, <<>>),
+                generation => 1,
+                retry => 5,
+                retry_interval => 100
+            },
+            WTrans = fun() ->
+                emqx_ds:trans(
+                    TxOpts,
+                    fun() ->
+                        emqx_ds:tx_write({[], ?ds_tx_ts_monotonic, <<"hello">>})
+                    end
+                )
+            end,
+            DTrans = fun() ->
+                emqx_ds:trans(
+                    TxOpts,
+                    fun() ->
+                        emqx_ds:tx_del_topic([])
+                    end
+                )
+            end,
+            Append = fun(N) ->
+                Ref = emqx_ds:dirty_append(TxOpts#{reply => true}, [
+                    {[], ?ds_tx_ts_monotonic, <<"hello">>}
+                 || _ <- lists:seq(1, N)
+                ]),
+                receive
+                    ?ds_tx_commit_reply(Ref, Reply) ->
+                        emqx_ds:dirty_append_outcome(Ref, Reply)
+                end
+            end,
+            %% 1. Start normally and write a sufficient volume of data to the DB to create some usage:
+            ?assertMatch(
+                {atomic, _, _},
+                WTrans()
+            ),
+            ?assertMatch(
+                {ok, _},
+                Append(1_000)
+            ),
+            %% TODO: this not entirely backend-agnostic:
+            emqx_ds_storage_layer:flush(DB),
+            ct:pal("DB group stats: ~p", [emqx_ds:db_group_stats(Group)]),
+            %% 2. Set quota to a very low value. This should cause all writes to fail:
+            %%
+            %% NOTE: currently OTX process checks quota during flush,
+            %% so we don't wait for alarms/events immediately:
```

Comment:
Perhaps this is not ideal from the user perspective: add a second alarm that is raised before quota is reached? That new one should be perhaps handled elsewhere, perhaps in the storage layer.